x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/filestructurefinder/TimestampFormatFinder.java

```diff
@@ -64,427 +113,1139 @@
-                    if (captures != null) {
-                        String preface = captures.getOrDefault(PREFACE, "").toString();
-                        String epilogue = captures.getOrDefault(EPILOGUE, "").toString();
-                        return makeTimestampMatch(candidate, index, preface, text.substring(preface.length(),
-                            text.length() - epilogue.length()), epilogue);
+                if (mustAdd) {
+                    if (errorOnMultiplePatterns && matchedFormats.isEmpty() == false) {
+                        throw new IllegalArgumentException("Multiple timestamp formats found ["
+                            + matchedFormats.get(0) + "] and [" + newFormat + "]");
                     }
+                    matchedFormats.add(newFormat);
                 }
+
+                matches.add(match);
+                cachedJavaTimestampFormats = null;
+                return;
             }
-            ++index;
         }
-        return null;
+
+        if (errorOnNoTimestamp) {
+            throw new IllegalArgumentException("No timestamp found in [" + text + "]");
+        }
     }
 
     /**
-     * Find the best timestamp format for matching an entire field value.
-     * @param text The value that the returned timestamp format must match in its entirety.
-     * @param timeoutChecker Will abort the operation if its timeout is exceeded.
-     * @return The timestamp format, or <code>null</code> if none matches.
+     * Where multiple timestamp formats have been found, select the "best" one, whose details
+     * will then be returned by methods such as {@link #getGrokPatternName} and
+     * {@link #getJavaTimestampFormats}.  If fewer than two timestamp formats have been found
+     * then this method does nothing.
+     */
+    public void selectBestMatch() {
+
+        if (matchedFormats.size() < 2) {
+            // Nothing to do
+            return;
+        }
+
+        int remainingMatches = matches.size();
+        double[] weights = new double[matchedFormats.size()];
+        for (TimestampMatch match : matches) {
+
+            for (int matchedFormatIndex = 0; matchedFormatIndex < matchedFormats.size(); ++matchedFormatIndex) {
+                if (matchedFormats.get(matchedFormatIndex).canMergeWith(match.timestampFormat)) {
+                    weights[matchedFormatIndex] += weightForMatch(match.preface);
+                    break;
+                }
+                ++matchedFormatIndex;
+            }
+
+            // The highest possible weight is 1, so if the difference between the two highest weights
+            // is less than the number of lines remaining then the leader cannot possibly be overtaken
+            if (findDifferenceBetweenTwoHighestWeights(weights) > --remainingMatches) {
+                break;
+            }
+        }
+
+        timeoutChecker.check("timestamp format determination");
+
+        double highestWeight = 0.0;
```

Comment:
This could be a `Tuple<Integer, Double> findHighestWeight(double[] weights)` method.